packages/react/src/components/SideMenu/DragHandleMenu/DefaultButtons/BlockColorsButton.tsx

```diff
@@ -1,16 +1,109 @@
-import { BlockSchema, PartialBlock } from "@blocknote/core";
+import {
+  Block,
+  BlockFromConfig,
+  BlockNoteEditor,
+  BlockSchema,
+  DefaultBlockSchema,
+  DefaultInlineContentSchema,
+  DefaultStyleSchema,
+  InlineContentSchema,
+  PropSchema,
+  StyleSchema,
+} from "@blocknote/core";
 import { Box, Menu } from "@mantine/core";
 import { ReactNode, useCallback, useRef, useState } from "react";
 import { HiChevronRight } from "react-icons/hi";
 
+import { useBlockNoteEditor } from "../../../../editor/BlockNoteContext";
 import { usePreventMenuOverflow } from "../../../../hooks/usePreventMenuOverflow";
-import { ColorPicker } from "../../../../components-shared/ColorPicker/ColorPicker";
-import type { DragHandleMenuProps } from "../DragHandleMenu";
+import { DragHandleMenuProps } from "../DragHandleMenuProps";
 import { DragHandleMenuItem } from "../DragHandleMenuItem";
+import { ColorPicker } from "../../../../components-shared/ColorPicker/ColorPicker";
+
+type BlockConfigWithColor<
+  Color extends "text" | "background",
+  T extends string = string,
+  PS extends PropSchema = PropSchema,
+  C extends "inline" | "table" | "none" = "inline" | "table" | "none"
+> = {
+  type: T;
+  propSchema: PS & Color extends "text"
+    ? {
+        textColor: {
+          default: string;
+        };
+      }
+    : {
+        backgroundColor: {
+          default: string;
+        };
+      };
+  content: C;
+};
 
-export const BlockColorsButton = <BSchema extends BlockSchema>(
-  props: DragHandleMenuProps<BSchema, any, any> & { children: ReactNode }
+function checkBlockTypeHasColor<
+  Color extends "text" | "background",
+  BSchema extends BlockSchema = DefaultBlockSchema,
+  I extends InlineContentSchema = DefaultInlineContentSchema,
+  S extends StyleSchema = DefaultStyleSchema
+>(color: Color, blockType: string, editor: BlockNoteEditor<BSchema, I, S>) {
+  return (
+    // Block type has color prop
+    blockType in editor.blockSchema &&
+    `${color}Color` in editor.blockSchema[blockType].propSchema &&
+    // Default textAlignment value is valid
+    !("values" in editor.blockSchema[blockType].propSchema[`${color}Color`])
+  );
+}
+
+function checkBlockHasColor<
+  Color extends "text" | "background",
+  BSchema extends BlockSchema = DefaultBlockSchema,
+  I extends InlineContentSchema = DefaultInlineContentSchema,
+  S extends StyleSchema = DefaultStyleSchema
+>(
+  color: Color,
+  block: Block<BSchema, I, S>,
+  editor: BlockNoteEditor<BSchema, I, S>
+): block is BlockFromConfig<BlockConfigWithColor<Color>, I, S> {
+  return checkBlockTypeHasColor(color, block.type, editor);
+}
+
+function checkBlockCanHaveColor<
```

Comment:
seems same as checkBlockCanHasColor?

Reply:
It is, but it asserts a type for the block rather than the editor. You unfortunately can't assert 2 types in a single guard, hence why there are 2 functions with the same logic but different typing.